sandbox/docker_entrypoint.sh

```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+set -Eeuo pipefail
+
+# Runs the command provided as arguments, but attempts to configure permissions first.
+
+important_read_dirs=("/grist" "/persist")
+important_write_dirs=("/persist")
+current_user_id=$(id -u)
+
+# We want to avoid running Grist as root if possible.
+# Try to setup permissions and de-elevate to a normal user.
+if [[ $current_user_id == 0 ]]; then
+  target_user=${GRIST_DOCKER_USER:-grist}
+  target_group=${GRIST_DOCKER_GROUP:-grist}
+
+  for dir in "${important_write_dirs[@]}"; do
+    # Make sure the target user owns everything that Grist needs write access to.
+    find "$dir" ! -user "$target_user" -exec chown "$target_user" "{}" +
+  done
+
+  # Restart as the target user, replacing the current process (replacement is needed for security).
```

Comment:
What security does `exec` give us? If we had a parent process, a security hole in `setpriv` would keep the root process running, and thus potentially exploitable?

Reply:
Exactly - exec throws away the parent process running as root.

Digging out the issue, it looks like it was a vuln mentioned in Gosu (an alternative to setpriv), concerning TTYs persisting across privilege boundaries:
https://github.com/tianon/gosu/issues/37

The `exec` gets rid of the parent shell, which removes vulnerabilities where malicious code can interact with it.